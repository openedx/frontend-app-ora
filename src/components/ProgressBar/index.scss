@import '@edx/paragon/scss/core/core';

.ora-progress-nav-group {
  max-width: $max-width-lg;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin: auto;
  .ora-progress-divider {
    position: absolute;
    max-width: $max-width-lg;
    width: 100%;
  }
  .ora-progress-nav {
    background-color: white;
    z-index: 5;
    font-size: 1.125rem;
    &.is-active {
      border-bottom: 2px solid black;
    }
  }
  .nav-icon {
    display: inline-block;
    margin-inline-end: 0.5rem;
    margin-left: -0.25rem;
    svg {
      display: inline-block;
    }
  }
  .number-icon {
    background-color: black;
    color: white;
    border-radius: 100%;
  }
  .active {
    border-bottom: 2px solid black;
  }
}

@mixin navbar-collapse-definition {
  width: 100%;
  flex-direction: column;
  align-items: start;
  border-top: 1px solid black;
  padding: 0 map-get($spacers, 3);

  .ora-progress-divider {
    display: none;
  }

  .ora-progress-nav {
    width: 100%;
    margin: 0 0.25rem;
  }
}

@include media-breakpoint-up(md) {
  .navbar-expand-lg .ora-progress-nav-group {
    @extend navbar-collapse-definition !optional;
  }
}

@include media-breakpoint-down(sm) {
  .navbar-expand-md .ora-progress-nav-group {
    @extend navbar-collapse-definition !optional;
  }
}
