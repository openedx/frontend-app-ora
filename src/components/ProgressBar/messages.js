import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  createSubmission: {
    id: 'frontend-app-ora.ProgressBar.createSubmission',
    defaultMessage: 'Create response',
    description: 'Create response progress indicator',
  },
  studentTraining: {
    id: 'frontend-app-ora.ProgressBar.studentTraining',
    defaultMessage: 'Practice grading',
    description: 'Student training step progress indicator',
  },
  selfAssess: {
    id: 'frontend-app-ora.ProgressBar.selfAssess',
    defaultMessage: 'Self grading',
    description: 'Self assessment step progress indicator',
  },
  peerAssess: {
    id: 'frontend-app-ora.ProgressBar.peerAssess',
    defaultMessage: 'Grade peers',
    description: 'Peer assessment step progress indicator',
  },
  myGrade: {
    id: 'frontend-app-ora.ProgressBar.myGrade',
    defaultMessage: 'My grade',
    description: 'My Grade step progress indicator',
  },
});

export default messages;
