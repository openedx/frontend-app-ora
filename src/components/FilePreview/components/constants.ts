import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from './FileRenderer/BaseRenderers';

import messages from './messages';

export const errorStatuses = {
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
} as const;

export const fileTypes = {
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
} as const;

export const errorMessages = {
  [errorStatuses.notFound]: messages.fileNotFoundError,
  [errorStatuses.serverError]: messages.unknownError,
};

/**
 * Config data
 */
export const renderers = {
  [fileTypes.pdf]: PDFRenderer,
  [fileTypes.jpg]: ImageRenderer,
  [fileTypes.jpeg]: ImageRenderer,
  [fileTypes.bmp]: ImageRenderer,
  [fileTypes.png]: ImageRenderer,
  [fileTypes.txt]: TXTRenderer,
  [fileTypes.gif]: ImageRenderer,
  [fileTypes.jfif]: ImageRenderer,
  [fileTypes.pjpeg]: ImageRenderer,
  [fileTypes.pjp]: ImageRenderer,
  [fileTypes.svg]: ImageRenderer,
};

export const supportedTypes = Object.keys(renderers);
