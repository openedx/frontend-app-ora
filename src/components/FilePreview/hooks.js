import { useKeyedState, StrictDict } from '@edx/react-unit-test-utils';

import {
  PDFRenderer,
  ImageRenderer,
  TXTRenderer,
} from 'components/FilePreview/BaseRenderers';

import messages from './messages';

export const ErrorStatuses = StrictDict({
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  conflict: 409,
  serverError: 500,
});

export const FileTypes = StrictDict({
  pdf: 'pdf',
  jpg: 'jpg',
  jpeg: 'jpeg',
  png: 'png',
  bmp: 'bmp',
  txt: 'txt',
  gif: 'gif',
  jfif: 'jfif',
  pjpeg: 'pjpeg',
  pjp: 'pjp',
  svg: 'svg',
});

/**
 * Config data
 */
export const RENDERERS = {
  [FileTypes.pdf]: PDFRenderer,
  [FileTypes.jpg]: ImageRenderer,
  [FileTypes.jpeg]: ImageRenderer,
  [FileTypes.bmp]: ImageRenderer,
  [FileTypes.png]: ImageRenderer,
  [FileTypes.txt]: TXTRenderer,
  [FileTypes.gif]: ImageRenderer,
  [FileTypes.jfif]: ImageRenderer,
  [FileTypes.pjpeg]: ImageRenderer,
  [FileTypes.pjp]: ImageRenderer,
  [FileTypes.svg]: ImageRenderer,
};

export const SUPPORTED_TYPES = Object.keys(RENDERERS);

export const ERROR_STATUSES = {
  [ErrorStatuses.notFound]: messages.fileNotFoundError,
  [ErrorStatuses.serverError]: messages.unknownError,
};

/**
 * Util methods and transforms
 */
export const getFileType = (fileName) => fileName.split('.').pop()?.toLowerCase();
export const isSupported = (file) => SUPPORTED_TYPES.includes(
  getFileType(file.name),
);

export const stateKeys = StrictDict({
  errorStatus: 'errorStatus',
  isLoading: 'isLoading',
});

/**
 * component hooks
 */
export const renderHooks = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useKeyedState(stateKeys.errorStatus, null);
  const [isLoading, setIsLoading] = useKeyedState(stateKeys.isLoading, true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    ERROR_STATUSES[errorStatus] || ERROR_STATUSES[ErrorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = RENDERERS[getFileType(file.name)];
  const rendererProps = {
    fileName: file.name,
    url: file.downloadUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
