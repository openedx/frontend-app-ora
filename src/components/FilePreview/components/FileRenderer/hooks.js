import { useState } from 'react';
import { errorStatuses, errorMessages, renderers } from '../constants';
import { getFileType } from '../utils';
import messages from './messages';

/**
 * component hooks
 */
export const useRenderData = ({
  file,
  formatMessage,
}) => {
  const [errorStatus, setErrorStatus] = useState(null);
  const [isLoading, setIsLoading] = useState(true);

  const setState = (newState) => {
    setErrorStatus(newState.errorStatus);
    setIsLoading(newState.isLoading);
  };

  const stopLoading = (status = null) => setState({ isLoading: false, errorStatus: status });

  const errorMessage = (
    errorMessages[errorStatus] || errorMessages[errorStatuses.serverError]
  );
  const errorAction = {
    id: 'retry',
    onClick: () => setState({ errorStatus: null, isLoading: true }),
    message: messages.retryButton,
  };
  const error = {
    headerMessage: errorMessage,
    children: formatMessage(errorMessage),
    actions: [errorAction],
  };

  const Renderer = renderers[getFileType(file.fileName)];
  const rendererProps = {
    fileName: file.fileName,
    url: file.fileUrl,
    onError: stopLoading,
    onSuccess: () => stopLoading(),
  };

  return {
    errorStatus,
    isLoading,
    error,
    Renderer,
    rendererProps,
  };
};
