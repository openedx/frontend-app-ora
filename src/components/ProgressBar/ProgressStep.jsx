import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { StrictDict } from '@edx/react-unit-test-utils';
import { Nav, Icon } from '@edx/paragon';
import {
  CheckCircle,
  Edit,
  Error,
  Highlight,
  Rule,
} from '@edx/paragon/icons';

import { stepNames } from 'constants/index';
import { useProgressStepData } from './hooks';

export const stepIcons = StrictDict({
  [stepNames.submission]: Edit,
  [stepNames.studentTraining]: Highlight,
  [stepNames.self]: Highlight,
  [stepNames.peer]: Highlight,
  [stepNames.done]: Rule,
});

const ProgressStep = ({
  step,
  canRevisit,
  label,
}) => {
  const {
    onClick,
    href,
    isActive,
    isEnabled,
    isComplete,
    isPastDue,
    myGrade,
  } = useProgressStepData({ step, canRevisit });
  let iconSrc = stepIcons[step];
  let subLabel = null;
  let colorClass = null;
  if (isPastDue) {
    colorClass = 'text-danger-500';
    iconSrc = Error;
    subLabel = 'Past due!';
  } else if (isComplete) {
    iconSrc = CheckCircle;
    if (step === stepNames.done && myGrade) {
      subLabel = `${myGrade.stepScore.earned} / ${myGrade.stepScore.possible}`;
    }
  }
  return (
    <Nav.Link
      {...(isEnabled && { onClick, href })}
      disabled={!isEnabled}
      className={classNames(
        'ora-progress-nav',
        'px-0',
        { 'is-active': isActive },
      )}
    >
      <Icon
        className={classNames('nav-icon', 'my-auto', colorClass)}
        src={iconSrc}
        {...subLabel && { style: { position: 'relative', bottom: '0.7rem' } }}
      />
      <div className="d-inline-block">
        {label}
        {subLabel && (
          <p data-testid="sublabel-test-id" className={classNames('x-small', 'm-0', colorClass)}>{subLabel}</p>
        )}
      </div>
    </Nav.Link>
  );
};
ProgressStep.propTypes = {
  label: PropTypes.node.isRequired,
  step: PropTypes.string.isRequired,
  canRevisit: PropTypes.bool.isRequired,
};

export default ProgressStep;
