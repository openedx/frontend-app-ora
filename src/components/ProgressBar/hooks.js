import { useParams } from 'react-router-dom';
import { useIsEmbedded, useViewStep } from 'hooks';
import {
  useEffectiveGrade,
  useGlobalState,
} from 'data/services/lms/hooks/selectors';
import {
  stepRoutes,
  stepStates,
} from 'data/services/lms/constants';

export const useProgressStepData = ({ step, canRevisit = false }) => {
  const { xblockId, courseId } = useParams();
  const isEmbedded = useIsEmbedded();
  const viewStep = useViewStep();
  const { stepState } = useGlobalState({ step });

  const href = `/${stepRoutes[step]}${isEmbedded ? '/embedded' : ''}/${courseId}/${xblockId}`;
  const isActive = viewStep === step;
  const isEnabled = (
    isActive
    || (stepState === stepStates.inProgress)
    || (canRevisit && stepState === stepStates.done)
  );
  const myGrade = useEffectiveGrade()?.stepScore;
  return {
    href,
    isEnabled,
    isActive,
    isComplete: stepState === stepStates.done,
    inProgress: stepState === stepStates.inProgress,
    isPastDue: stepState === stepStates.closed,
    myGrade,
    // myGrade: { earned: 8, possible: 10 },
    // isPastDue: step === 'self',
  };
};

export default { useProgressStepData };
