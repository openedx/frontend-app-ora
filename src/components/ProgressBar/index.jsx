import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Navbar, Icon } from '@edx/paragon';
import { ArrowDropUpDown, ArrowForwardIos } from '@edx/paragon/icons';

import {
  useAssessmentStepOrder,
  useGlobalState,
  useHasReceivedFinalGrade,
  useIsPageDataLoaded,
} from 'hooks/app';
import { stepNames } from 'constants/index';

import { useViewStep } from 'hooks/routing';
import { isXblockStep } from 'utils';

import ProgressStep from './ProgressStep';

import messages from './messages';
import './index.scss';

export const stepLabels = {
  [stepNames.submission]: messages.createSubmission,
  [stepNames.peer]: messages.peerAssess,
  [stepNames.studentTraining]: messages.studentTraining,
  [stepNames.self]: messages.selfAssess,
  [stepNames.done]: messages.myGrade,
};

export const stepCanRevisit = {
  [stepNames.submission]: true,
  [stepNames.peer]: true,
  [stepNames.studentTraining]: false,
  [stepNames.self]: false,
  [stepNames.done]: true,
};

const ProgressBar = ({ className }) => {
  const isLoaded = useIsPageDataLoaded();
  const hasReceivedFinalGrade = useHasReceivedFinalGrade();

  const activeStep = useViewStep();
  const { activeStepName } = useGlobalState();
  const isXblock = isXblockStep(activeStep);

  const stepOrders = [
    stepNames.submission,
    ...useAssessmentStepOrder(),
    stepNames.done,
  ].filter(step => step !== stepNames.staff);
  const { formatMessage } = useIntl();

  if (!isLoaded) {
    return null;
  }

  const stepEl = (curStep) => (
    <ProgressStep
      step={curStep}
      key={curStep}
      label={formatMessage(stepLabels[curStep])}
      canRevisit={
          (curStep === 'done' && hasReceivedFinalGrade)
          || stepCanRevisit[curStep]
        }
    />
  );

  let activeStepTitle = isXblock ? activeStepName : activeStep;
  if (activeStepTitle === stepNames.staff) {
    activeStepTitle = stepNames.submission;
  }
  const stepIndex = stepOrders.indexOf(activeStepTitle) + 1;
  const totalSteps = stepOrders.length;

  return (
    <Navbar className={classNames('px-0', className)} expand={stepOrders.length > 4 ? 'lg' : 'md'}>
      <Navbar.Toggle className="w-100 border-primary">
        <div className="d-flex justify-content-between m-0 h3">
          <span className="d-flex">
            <Icon src={ArrowForwardIos} className="mr-2" />
            {formatMessage(stepLabels[activeStepTitle])}
            {formatMessage(messages.step, {
              stepIndex,
              totalSteps,
            })}
          </span>
          <Icon src={ArrowDropUpDown} />
        </div>
      </Navbar.Toggle>
      <Navbar.Collapse className="ora-progress-nav-group bg-white">
        <hr className="ora-progress-divider" />
        {stepOrders.map(stepEl)}
      </Navbar.Collapse>
    </Navbar>
  );
};
ProgressBar.defaultProps = {
  className: '',
};
ProgressBar.propTypes = {
  className: PropTypes.string,
};

export default ProgressBar;
