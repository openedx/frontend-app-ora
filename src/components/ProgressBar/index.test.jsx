import { shallow } from '@edx/react-unit-test-utils';

import {
  useAssessmentStepOrder,
  useGlobalState,
  useHasReceivedFinalGrade,
  useIsPageDataLoaded,
} from 'hooks/app';
import { stepNames } from 'constants/index';
import { useViewStep } from 'hooks/routing';
import { isXblockStep } from 'utils';

import ProgressBar from './index';

jest.mock('hooks/app', () => ({
  useAssessmentStepOrder: jest.fn(),
  useGlobalState: jest.fn(),
  useHasReceivedFinalGrade: jest.fn(),
  useIsPageDataLoaded: jest.fn(),
}));
jest.mock('hooks/routing', () => ({
  useViewStep: jest.fn(),
}));
jest.mock('utils', () => ({
  isXblockStep: jest.fn(),
}));
jest.mock('./ProgressStep', () => 'ProgressStep');

describe('<ProgressBar />', () => {
  const props = {
    className: 'test-class',
  };

  beforeEach(() => {
    useIsPageDataLoaded.mockReturnValue(true);
    useHasReceivedFinalGrade.mockReturnValue(false);
    useGlobalState.mockReturnValue({ activeStepName: stepNames.submission });
    useAssessmentStepOrder.mockReturnValue([]);
    useViewStep.mockReturnValue(stepNames.submission);
    isXblockStep.mockReturnValue(false);
  });

  it('renders null when page data is not loaded', () => {
    useIsPageDataLoaded.mockReturnValueOnce(false);
    const wrapper = shallow(<ProgressBar {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();

    expect(wrapper.isEmptyRender()).toBe(true);
  });

  it('renders at least 2 steps: submission and done', () => {
    const wrapper = shallow(<ProgressBar {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();
    expect(wrapper.instance.findByType('ProgressStep')).toHaveLength(2);
  });

  it('renders all steps', () => {
    isXblockStep.mockReturnValueOnce(true);
    useAssessmentStepOrder.mockReturnValueOnce([
      stepNames.studentTraining,
      stepNames.self,
      stepNames.peer,
    ]);
    const wrapper = shallow(<ProgressBar {...props} />);
    expect(wrapper.snapshot).toMatchSnapshot();
    expect(wrapper.instance.findByType('ProgressStep')).toHaveLength(5);
  });
});
