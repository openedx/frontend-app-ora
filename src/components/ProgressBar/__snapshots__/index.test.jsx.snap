// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProgressBar /> renders all steps 1`] = `
<Navbar
  className="px-0 test-class"
  expand="lg"
>
  <Navbar.Toggle
    className="w-100 border-primary"
  >
    <div
      className="d-flex justify-content-between m-0 h3"
    >
      <span
        className="d-flex"
      >
        <Icon
          className="mr-2"
        />
        Create response
         (Step 1/5)
      </span>
      <Icon />
    </div>
  </Navbar.Toggle>
  <Navbar.Collapse
    className="ora-progress-nav-group bg-white"
  >
    <hr
      className="ora-progress-divider"
    />
    <ProgressStep
      canRevisit={true}
      key="submission"
      label="Create response"
      step="submission"
    />
    <ProgressStep
      canRevisit={false}
      key="studentTraining"
      label="Practice grading"
      step="studentTraining"
    />
    <ProgressStep
      canRevisit={false}
      key="self"
      label="Self grading"
      step="self"
    />
    <ProgressStep
      canRevisit={true}
      key="peer"
      label="Peer grading"
      step="peer"
    />
    <ProgressStep
      canRevisit={true}
      key="done"
      label="My grade"
      step="done"
    />
  </Navbar.Collapse>
</Navbar>
`;

exports[`<ProgressBar /> renders at least 2 steps: submission and done 1`] = `
<Navbar
  className="px-0 test-class"
  expand="md"
>
  <Navbar.Toggle
    className="w-100 border-primary"
  >
    <div
      className="d-flex justify-content-between m-0 h3"
    >
      <span
        className="d-flex"
      >
        <Icon
          className="mr-2"
        />
        Create response
         (Step 1/2)
      </span>
      <Icon />
    </div>
  </Navbar.Toggle>
  <Navbar.Collapse
    className="ora-progress-nav-group bg-white"
  >
    <hr
      className="ora-progress-divider"
    />
    <ProgressStep
      canRevisit={true}
      key="submission"
      label="Create response"
      step="submission"
    />
    <ProgressStep
      canRevisit={true}
      key="done"
      label="My grade"
      step="done"
    />
  </Navbar.Collapse>
</Navbar>
`;

exports[`<ProgressBar /> renders null when page data is not loaded 1`] = `null`;
